tools/codegen/gen.py

```diff
@@ -259,38 +261,29 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
 
                 has_tensor_options = any(isinstance(a.argument, TensorOptionsArguments) for a in args)
 
-                if local.use_c10_dispatcher() == UseC10Dispatcher.full:
-                    cuda_guard_from_tensor_options = """\
-    const DeviceGuard device_guard(device_or_default(device));
-"""
-                else:
-                    assert local.use_c10_dispatcher() in [UseC10Dispatcher.with_codegenerated_unboxing_wrapper,
-                                                          UseC10Dispatcher.hacky_wrapper_for_legacy_signatures]
-                    cuda_guard_from_tensor_options = """\
-    const DeviceGuard device_guard(options.device());
-"""
-
-                # TODO: There is probably a simpler version of this that
-                # works just as well.
-                if f.device_guard and is_generic_dispatch_key(self.dispatch_key) and has_tensor_options:
-                    cuda_guard = cuda_guard_from_tensor_options
-                elif f.device_guard and is_cuda_dispatch_key(self.dispatch_key) and has_tensor_options:
-                    cuda_guard = f"""\
-    globalContext().lazyInitCUDA();
-    {cuda_guard_from_tensor_options}
-"""
+                # Initialization happens on construction of CUDA tensors, so
+                # that all methods/functions that operates on CUDA tensors can
+                # assume initialization has already occurred
+                if is_cuda_dispatch_key(self.dispatch_key) and has_tensor_options:
+                    init_cuda = "globalContext().lazyInitCUDA();"
+
+                if f.device_guard and has_tensor_options:
+                    if local.use_c10_dispatcher() is UseC10Dispatcher.full:
+                        cuda_guard = "const DeviceGuard device_guard(device_or_default(device));"
+                    else:
+                        assert local.use_c10_dispatcher() in [UseC10Dispatcher.with_codegenerated_unboxing_wrapper,
+                                                              UseC10Dispatcher.hacky_wrapper_for_legacy_signatures]
```

Comment:
I know it's not new code, but is this assert necessary? It's a 3-place enum, right?

Reply:
We litigated this in https://github.com/pytorch/pytorch/pull/45742 (see https://github.com/pytorch/pytorch/pull/45742#issuecomment-703796733) and smessmer did add the helper functions, but it looks like a bunch of direct tests against the individual enums crept back in (and I approved them, oops!)